File: net_growth.m
function net = net_growth(method,N_max,NR,ER,t)
    
    % Initial network is a network of two nodes connected by a single edge
    N = 2;                      
    net = zeros(N,N);
    net(2,1) = 1;
    net(1,2) = 1;

    % Network generation

    % for each time point:
    for time = 1:t
        
        % If their are no nodes, stop growth
        if isempty(net)
            break
        end

    % -----
    % NODES
    % -----

        % First of all work out what the change in network size is
        A = N/N_max;                % node fraction max nodes
        A = NR*A*(1-A);             % work out change as fraction
        N_new = round(A*N_max);     % updated number of nodes 
        N_dif = N_new - N;          % difference in nodes

        % Add/remove nodes

        if N_dif > 0 % add new nodes
            row = zeros(length(net),N_dif);         
            col = zeros(N_dif,length(net)+N_dif);   
            net = [net row]; 
            net = [net; col];                        
            % index of new nodes
            new_nodes = N:N_new;        
        elseif N_dif< 0
            if abs(N_dif)>length(net)
                N_dif = length(net);
            end
            % randomly select number of nodes = to difference and remove row and column from matrix
            kill = randperm(length(net),abs(N_dif));
            % remove rows/cols
            net(kill,:) = []; 
            net(:,kill) = [];
        end 

    % -----
    % EDGES
    % -----

        if method == 0
            % if nodes added
            if N_dif > 0
            % for each new node (new_nodes)
                for new = 1:length(new_nodes)
                    E_new = zeros(length(net),1);
                    for E = 1:length(E_new)
                        E_new(E) = round(rand);
                    end
                    net(:,new_nodes(new)) = E_new;
                    net(new_nodes(new),:) = E_new;
                end
            end

        elseif method == 1

            % update network edges - in this version random chance of connection to
            % any existing node
            % determine the number of edges to add/remove using equation
            % based on current no. of nodes
            % get current fraction of capacity of EDGES
            cur_E = nnz(tril(net,-1)); % current number of edges, NB only take half, as symetric matrix
            max_E = round((((length(net)*length(net))-length(net))/2)); % maximum possible number of edges - (n squared - n) over 2
            fract_E = cur_E/max_E; % current fraction of edges
            new_E = ER*fract_E * (1-fract_E); % new fraction of edges
            E_new = round(new_E*max_E); % updated number of edges
            E_dif = E_new - cur_E; % difference in edges
            if E_dif > 0
                New_edges = nan(1,E_dif);
                edge_loop = 0;
                while edge_loop <= E_dif
                    edge_loop = edge_loop + 1;
                    % draw a number between 1 and length of nodes (a)
                    a = randi(length(net));
                    % draw a second one (b)
                    b = randi(length(net));
                    if a == b
                        continue
                    end
                    if net(a,b) == 1
                        continue
                    end
                    net(a,b) = 1;
                    net(b,a) = 1;
                    
                end
            elseif E_dif < 0
                New_edges = nan(1,abs(E_dif));
                edge_loop = 0;
                while edge_loop <= abs(E_dif)  
                    edge_loop = edge_loop + 1;
                    % draw a number between 1 and length of nodes (a)
                    a = randi(length(net));
                    % draw a second one (b)
                    b = randi(length(net));
                    if a == b
                        continue
                    end
                    if net(a,b) == 0
                        continue
                    end
                    net(a,b) = 0;
                    net(b,a) = 0;
                    
                end           
            end
        end
    % --------
    % CLEAN UP
    % --------

        % make sure/ check diagonal == 0
        for i = 1: length(net)
            net(i,i) = 0;
        end

        % remove any nodes with no edges
        remove = [];
        for i = 1:length(net)
            if sum(net(i,:)) == 0
                if sum(net(:,i)) == 0
                    % creat a vector of indicies where this is true
                    remove = [remove i];
                end
            end
        end
        if isempty(remove) == 0
            net(remove,:) = [];
            net(:,remove) = [];
        end
        empty(time) = length(remove);

        N = N_new;
    end
end

File: net_growth.py
def logistic(growth_rate = 2,carry_capacity = 0.3):
    return growth_rate*carry_capacity*(1-carry_capacity)

def logistic_net_binary(Nr=2,Er=2,Max_n=100,Iterations=50):
    adj = np.array([[0,1],[1,0]])
    growth = []
    cost = []
    for i in tqdm(range(Iterations)):
        growth.append(len(adj))
        
        # Number of nodes to add
        
        # new_n is the change in the numeber of nodes (+ve = increase, -ve decrease), so we convert the number of nodex to fraciton of carry capacity, work out xt+1, round to whole nodes, and subtract the current no. of nodes
        new_n = round(Max_n * logistic(Nr,len(adj)/Max_n)) - len(adj)

        # add or remove nodes
        if new_n > 0:
            n = len(adj)+new_n
            b = np.zeros((n,n))
            b[:len(adj),:len(adj)] = adj
            adj = np.copy(b)
        elif new_n < 0:
            remove = random.sample(range(0,len(adj)),abs(new_n))
            adj = np.delete(adj,remove,0)
            adj = np.delete(adj,remove,1)
            
        # stop if no network
        if len(adj) == 0:
            break

        # work out how many edges to add/remove    

        # get the current number of edges
        curr_e = np.count_nonzero(np.tril(adj))
        # get the current maximum number of edges
        max_e = (len(adj))*(len(adj)-1)/2
        # based on this, get the number of edges to add/remove
        new_e = round(max_e * logistic(Er, curr_e/max_e)) - curr_e
        


        # add/ remove edges

        # Add
        if new_e > 0:
            for e in range(new_e):
                row = random.randint(0,len(adj)-1)
                # get a column index
                col = random.randint(0,len(adj)-1)
                # if row == col, repeat until it doesnt
                while row == col:
                    row = random.randint(0,len(adj)-1)
                    col = random.randint(0,len(adj)-1)

                while adj[row,col] > 0:
                    row = random.randint(0,len(adj)-1)
                    col = random.randint(0,len(adj)-1)
                    while row == col:
                        row = random.randint(0,len(adj)-1)
                        col = random.randint(0,len(adj)-1)
                adj[row,col] = 1
                adj[col,row] = 1    

        # Remove
        elif new_e < 0:
            for e in range(abs(new_e)):
                row = random.randint(0,len(adj)-1)
                col = random.randint(0,len(adj)-1)
                while row == col:
                    row = random.randint(0,len(adj)-1)
                    col = random.randint(0,len(adj)-1)
                while adj[row,col] == 0:
                    row = random.randint(0,len(adj)-1)
                    col = random.randint(0,len(adj)-1)
                    while row == col:
                        row = random.randint(0,len(adj)-1)
                        col = random.randint(0,len(adj)-1)
                adj[row,col] = 0
                adj[col,row] = 0 

            # clean up - remove any unconnected nodes (a node dies if it is not part of a network)
        to_kill = [x for x in range(len(np.sum(adj,0))) if np.sum(adj,0)[x] == 0]
        if len(to_kill)>0:
            adj = np.delete(adj,to_kill,0)
            adj = np.delete(adj,to_kill,1) 
            
        cost.append(curr_e/max_e)
    
    return adj,growth,cost
